### PR TITLE
find if is there  any overlap range 

### LINKED ISSUE #2989: Store: Fix check for existing overlapping replica
`Store.addReplicaInternal` and `CanApplySnapshot` use `rangesByKey.Has()` to detect if a range already exists, but this only looks at a single key so it may be incorrect when there have been multiple splits or merges. This check should be replaced with one that looks at the start and end points of the replica.

See #2944

### PR DESCRIPTION
**Pull request by veteranlu**
cc @bdarnell look up `replicasByKey`  to get overlap range.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3171)

### CODE REVIEW

File: storage/store.go [hunk 1]
```diff
@@ -1822,3 +1822,21 @@ func (s *Store) PublishStatus() error {
...
+// hasReplicaExist used to judgement if "rngDesc" has overlap with ranges
+// in "replicasByKey" btree, it is not thread saft.
+func (s *Store) hasOverlapReplica(rngDesc *roachpb.RangeDescriptor) bool {
+	var rng *Replica
+	s.replicasByKey.AscendGreaterOrEqual((rangeBTreeKey)(rngDesc.StartKey.Next()), func(i btree.Item) bool {
+		rng = i.(*Replica)
+		return false
+	})
+
+	if rng != nil {
```
Comment (tschottdorf) [inline storage/store.go]: everything below this line is equal to

`return rng != nil && rng.Desc().StartKey.Less(rngDesc.EndKey)`

File: storage/store.go [hunk 2]
```diff
@@ -1822,3 +1822,21 @@ func (s *Store) PublishStatus() error {
 func (s *Store) SetRangeRetryOptions(ro retry.Options) {
 	s.ctx.RangeRetryOptions = ro
 }
+
+// hasReplicaExist used to judgement if "rngDesc" has overlap with ranges
+// in "replicasByKey" btree, it is not thread saft.
+func (s *Store) hasOverlapReplica(rngDesc *roachpb.RangeDescriptor) bool {
```
Comment (tschottdorf) [inline storage/store.go]: rename this to `hasOverlappingReplicaLocked`.

Comment (tschottdorf) [inline storage/store.go]: >
// hasOverlappingReplicaLocked returns true if a Replica overlapping the given descriptor is present on the Store.

Comment (bdarnell) [inline storage/store.go]: Also move this method next to the similar `Store.LookupReplica`.

**Comment by bdarnell** [comment]
Can you add a test for this method? Also, you or your editor have set the executable bit on the source file; please undo that.

**Comment by veteranlu** [comment]
@bdarnell sure

**Comment by veteranlu** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

_[storage/store.go, line 1828 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3171#-K3SvQag780li-98WirN-r1-1828:-K3XOq_Z-RctGGF-Rsh4:-9457221) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bcee2ac4e330ea5e8d6c19a4a3e4907aa0f7c31/storage/store.go#L1828)):_
@tschottdorf Is is properly named `hasOverlappingReplicaUnlocked`, since which access `replicasByKey` without lock?

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3171)_

**Comment by tschottdorf** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

_[storage/store.go, line 1828 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3171#-K3SvQag780li-98WirN-r1-1828:-K3XR-GPJmGawqngBAug:-153604922) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bcee2ac4e330ea5e8d6c19a4a3e4907aa0f7c31/storage/store.go#L1828)):_
calling something `xxxLocked` signifies that if you call it, you must be holding the lock.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3171)_

**Comment by tschottdorf** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3171)_

**Comment by veteranlu** [comment]
@tschottdorf can u help me review the code? I am pleasure to know if there is anything to improve it.

**Comment by tschottdorf** [comment]
Looks good! Could you squash the two commits into one? Thanks!

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_[storage/store_test.go, line 503 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3171#-K3bj0mJGmWA1lY6yio0:-K3bj0mJGmWA1lY6yio1:-125316798) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d550d8679af9ceeea40d72b9633ec7c13802ce6/storage/store_test.go#L503)):_
can you add `{roachpb.RKey("e"), roachpb.RKey("e\x00"), true}`?

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3171#-:-K3bjGin0epIi4BjMU27:1850465174)_

**Comment by bdarnell** [comment]
LGTM

**Pull request merged**